pkg/federation/sdlmerge/remove_duplicate_fielded_shared_types.go

```diff
@@ -0,0 +1,128 @@
+package sdlmerge
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astvisitor"
+	"github.com/jensneuse/graphql-go-tools/pkg/operationreport"
+)
+
+type removeDuplicateFieldedSharedTypesVisitor struct {
+	*astvisitor.Walker
+	document          *ast.Document
+	entitySet         map[string]bool
+	sharedTypeSet     map[string]FieldedSharedType
+	rootNodesToRemove []ast.Node
+	lastInputRef      int
+	lastInterfaceRef  int
+	lastObjectRef     int
+}
+
+func newRemoveDuplicateFieldedSharedTypesVisitor() *removeDuplicateFieldedSharedTypesVisitor {
+	return &removeDuplicateFieldedSharedTypesVisitor{
+		nil,
+		nil,
+		make(map[string]bool),
+		make(map[string]FieldedSharedType),
+		nil,
+		ast.InvalidRef,
+		ast.InvalidRef,
+		ast.InvalidRef,
+	}
+}
+
+func (r *removeDuplicateFieldedSharedTypesVisitor) Register(walker *astvisitor.Walker) {
+	r.Walker = walker
+	walker.RegisterEnterDocumentVisitor(r)
+	walker.RegisterEnterInputObjectTypeDefinitionVisitor(r)
+	walker.RegisterEnterInterfaceTypeDefinitionVisitor(r)
+	walker.RegisterEnterObjectTypeDefinitionVisitor(r)
+	walker.RegisterLeaveDocumentVisitor(r)
+}
+
+func (r *removeDuplicateFieldedSharedTypesVisitor) EnterDocument(operation, _ *ast.Document) {
+	r.document = operation
+}
+
+func (r *removeDuplicateFieldedSharedTypesVisitor) EnterInputObjectTypeDefinition(ref int) {
+	if ref <= r.lastInputRef {
+		return
+	}
+	document := r.document
+	name := document.InputObjectTypeDefinitionNameString(ref)
+	inputType := document.InputObjectTypeDefinitions[ref]
+	r.checkForDuplicateEntity(inputType.HasDirectives, inputType.Directives.Refs, name)
```

Comment:
Each entry into a fielded type now calls this method (defined below). The method checks whether the fielded type is an entity, and if it is, adds it to the entity set. If the entity name already exists in the set, we know it's duplicated, and the federation is immediately invalid.

Reply:
only object types and interfaces can be entitites (a.k.a. have a `@key` directive)

See: https://www.apollographql.com/docs/federation/v1/federation-spec
```graphql
directive @key(fields: _FieldSet!) repeatable on OBJECT | INTERFACE
```

Reply:
Of course. And that's why there is not visitor. Sorry.

If I remove this, are you happy with the implementation?